### ISSUE TITLE
special-case try/except ImportError if all imports resolve

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#2096]
> I encountered a case of this with library `mypy_boto3_s3` ([pypi](https://pypi.org/project/mypy-boto3-s3/), [github generator code](https://github.com/youtype/mypy_boto3_builder)). It uses a `try / except ImportError` pattern, I think for Python backwards compatibility?
> 
> ```python
> # mypy_boto3_s3/__init__.py
> try:
>     from .service_resource import S3ServiceResource
> except ImportError:
>     from builtins import object as S3ServiceResource  # type: ignore[assignment]
> ```
> 
> Unfortunately, ty doesn't handle this as gracefully as mypy. Mypy seems to recognize my `var: mypy_boto3_s3.S3ServiceResource` as an instance of `mypy_boto3_s3.service_resource.S3ServiceResource`, but ty instead widens it to `object`, and then it doesn't type check my code correctly.
> 
> I have no suggestion on how to handle this correctly, I am simply reporting that I encountered this issue in the wild.
> 
> Here's a basic reproduction case:
> 
> ```python
> # mypy_boto3_s3_test.py
> from mypy_boto3_s3 import S3ServiceResource
> 
> def f() -> S3ServiceResource:
>     """Fake making an S3ServiceResource"""
>     return None  # type: ignore
> 
> s = f()
> s.get_available_subresources()  # ty thinks this is an error
> 
> ```
> 
> ```
> $ uvx ty version
> ty 0.0.2
> 
> $ uvx ty check mypy_boto3_s3_test.py 
> error[unresolved-attribute]: Object of type `object` has no attribute `get_available_subresources`
>  --> mypy_boto3_s3_test.py:9:1
>   |
> 8 | s = f()
> 9 | s.get_available_subresources()  # ty thinks this is an error
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   |
> info: rule `unresolved-attribute` is enabled by default
> 
> Found 1 diagnostic
> ``` 

 _Originally posted by @leon-sony in [#1585](https://github.com/astral-sh/ty/issues/1585#issuecomment-3667714012)_

**Comment by carljm** [comment]
I think ty is more capable than mypy in this case, not less -- but since this package was designed around mypy's behavior, the results with this package are worse.

Mypy simply doesn't allow the double definition of `S3ServiceResource`, which is why the `type: ignore` on the second definition is required. And in this error case, mypy simply ignores the second definition and prefers the first. This is a bit arbitrary, but it happens to be the mypy behavior, so it's the behavior that the authors of this code generator rely on.

ty supports multiple definitions of the same name, and infers a union type, which is correct: according to the code, at runtime, the name could be either a real `S3ServiceResource`, or it could be the fallback of `object`. This union type simplifies to `object` (because all types are subtypes of `object`), and thus the behavior you see.

The only thing I could really see us doing here is adding some special handling for `try/except` when the `try` block consists entirely of import statements and the `except` is an `except ImportError`, and all the imports resolve successfully under the current module resolution settings. In this case, we could consider the `except` block to be unreachable code.

**Comment by carljm** [comment]
This issue is sort of just the reincarnation of #309, except this is an example of a case where disabling `possibly-missing-import` rule doesn't help.

**Comment by MichaReiser** [comment]
Related case where we run into this on the server side https://github.com/astral-sh/ty/issues/2249#issuecomment-3695877740

**Comment by MichaReiser** [comment]
I'll move this to stable as it also impacts the LSP

**Comment by miketheman** [comment]
Another real-world case: `factory_boy` uses this pattern in `factory/__init__.py` to conditionally import framework-specific submodules.
https://github.com/FactoryBoy/factory_boy/blob/ae9f2f4650afef0bc9b0925de97f618603233ff8/factory/__init__.py#L57-L72

The submodule `factory/django.py` exists on disk and is importable when Django is installed.

With ty 0.0.17, any usage of `factory.django` triggers `possibly-missing-attribute`:

```shellsession
$ ty check
warning[possibly-missing-attribute]: Member `django` may be missing on module `factory`
  --> tests/conftest.py:9:19
   |
 9 | class UserFactory(factory.django.DjangoModelFactory):
   |                   ^^^^^^^^^^^^^^
10 |     class Meta:
11 |         model = User
   |
info: rule `possibly-missing-attribute` is enabled by default
```

Do you need any further reproducer code for this issue?

**Comment by carljm** [comment]
@miketheman Thanks for the additional real-world example! I think it's pretty clear, no further reproducer code needed.

**Comment by mikelei8291** [comment]
I've also stumbled upon this issue when using the [evdev](https://github.com/gvalkov/python-evdev/blob/a47b5b5a6f79bde6823095d1105501856338aed7/src/evdev/device.py#L7-L10) library, and ty was not resolving methods defined in the parent class [`EventIO`](https://github.com/gvalkov/python-evdev/blob/a47b5b5a6f79bde6823095d1105501856338aed7/src/evdev/eventio_async.py#L56) of the [`InputDevice`](https://github.com/gvalkov/python-evdev/blob/a47b5b5a6f79bde6823095d1105501856338aed7/src/evdev/device.py#L105) class.

However, the good old [Jedi](https://github.com/davidhalter/jedi) and its successor [Zuban](https://github.com/zubanls/zuban) are doing it just fine (Zuban has some other minor issues on the imported class but the method resolving part is fine). I tested both locally. For Zuban you will need to use it with a compatible editor to see the effect, but for Jedi you can reproduce with the following code:

```python
import jedi
source = """
from evdev import InputDevice
d = InputDevice("/dev/input/event3")
d.async_
"""
script = jedi.Script(source, path="example.py")
print(script.complete(4, len("d.async_")))
# [<Completion: async_read>, <Completion: async_read_loop>, <Completion: async_read_one>]
```

**Comment by carljm** [comment]
@mikelei8291 thanks for the additional library case, useful so we can check that a fix here covers all the mentioned libraries.

I think my [comment above](https://github.com/astral-sh/ty/issues/2096#issuecomment-3673003509) about mypy applies equally to zuban in this case.
